torch/distributions/mixture_same_family.py

```diff
@@ -0,0 +1,181 @@
+    has_rsample = False
+
+    def __init__(self,
+                 mixture_distribution,
+                 component_distribution,
+                 validate_args=None):
+        self._mixture_distribution = mixture_distribution
+        self._component_distribution = component_distribution
+
+        if not isinstance(self._mixture_distribution, Categorical):
+            raise ValueError(" The Mixture distribution needs to be an "
+                             " instance of torch.distribtutions.Categorical")
+
+        if not isinstance(self._component_distribution, Distribution):
+            raise ValueError("The Component distribution need to be an "
+                             "instance of torch.distributions.Distribution")
+
+        # Check that batch size matches
+        mdbs = self._mixture_distribution.batch_shape
+        cdbs = self._component_distribution.batch_shape[:-1]
+        if len(mdbs) != 0 and mdbs != cdbs:
+            raise ValueError("`mixture_distribution.batch_shape` ({0}) is not "
+                             "compatible with `component_distribution."
+                             "batch_shape`({1})".format(mdbs, cdbs))
+
+        # Check that the number of mixture component matches
+        km = self._mixture_distribution.logits.shape[-1]
+        kc = self._component_distribution.batch_shape[-1]
+        if km is not None and kc is not None and km != kc:
+            raise ValueError("`mixture_distribution component` ({0}) does not"
+                             " equal `component_distribution.batch_shape[-1]`"
+                             " ({1})".format(km, kc))
+        self._num_component = km
+
+        event_shape = self._component_distribution.event_shape
+        self._event_ndims = len(event_shape)
+        super(MixtureSameFamily, self).__init__(batch_shape=cdbs,
+                                                event_shape=event_shape,
+                                                validate_args=validate_args)
+
+    def expand(self, batch_shape, _instance=None):
+        batch_shape = torch.Size(batch_shape)
+        batch_shape_comp = batch_shape + (self._num_component,)
+        new = self._get_checked_instance(MixtureSameFamily, _instance)
+        new._component_distribution = \
+            self._component_distribution.expand(batch_shape_comp)
+        new._mixture_distribution = \
+            self._mixture_distribution.expand(batch_shape)
+        new._num_component = self._num_component
+        new._event_ndims = self._event_ndims
+        event_shape = new._component_distribution.event_shape
+        super(MixtureSameFamily, new).__init__(batch_shape=batch_shape,
+                                               event_shape=event_shape,
+                                               validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    @constraints.dependent_property
+    def support(self):
+        return self._component_distribution.support
```

Comment:
FYI this will be wrong if the support is batched (e.g. `Uniform(lb, ub)` whose endpoints vary). However this is relatively rare and I see no immediate fix, so I think it's fine to leave it in. Maybe add a comment?
```py
# FIXME this may have the wrong shape when support contains batched parameters
```